crates/downloader/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+
+            result.push((blob_header, bundle));
+        }
+    }
+
+    Ok(result)
+}
+
+pub struct Downloader {
+    /// Ethereum provider client.
+    provider: RootProvider<Http<reqwest::Client>>,
+    /// HTTP client for downloading blobs.
+    http: reqwest::Client,
+    /// API URL for the beacon.
+    beacon_url: String,
+    /// Contract to monitor for new blobs.
+    target_contract: Address,
+    /// Current block number.
+    current_block: u64,
+    /// Cached block at `current_block + 1`.
+    /// Used to avoid fetching the same block multiple times.
+    peek_block: Option<Block>,
+}
+impl Downloader {
+    pub fn new(config: Config) -> anyhow::Result<Self> {
+        let rpc_url = config.ethereum_rpc_url.parse()?;
+        let provider = ProviderBuilder::new().on_http(rpc_url);
+        let http = reqwest::Client::new();
+        let target_contract: Address = config.blob_contract.parse()?;
+        Ok(Self {
+            provider,
+            http,
+            beacon_url: config.beacon_rpc_url,
+            target_contract,
+            current_block: config.start_block,
+            peek_block: None,
+        })
+    }
+
+    async fn download_block(&mut self, number: u64) -> anyhow::Result<Block> {
+        if self.current_block + 1 == number {
+            if let Some(block) = self.peek_block.clone() {
+                return Ok(block);
+            }
+        }
+
+        let block = self
+            .provider
+            .get_block_by_number(BlockNumberOrTag::Number(number), true)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("block not found"))?;
+
+        if self.current_block + 1 == number {
+            self.peek_block = Some(block.clone());
+        }
+
+        Ok(block)
+    }
+
+    async fn get_sidecars_for_current_block(
+        &mut self,
+    ) -> anyhow::Result<Option<BlobSidecarResponse>> {
+        let next_block = self.download_block(self.current_block + 1).await?;
+
+        let Some(next_pbbr) = next_block.header.parent_beacon_block_root else {
+            // No parent beacon block root, which means no blobs either.
+            return Ok(None);
+        };
+
+        let url = format!(
+            "{}/eth/v1/beacon/blob_sidecars/{}",
+            self.beacon_url, next_pbbr
+        );
+
+        Ok(Some(self.http.get(&url).send().await?.json().await?))
+    }
+
+    /// Fetches bundles and advances to the next block if successful.
+    pub async fn next_block_bundles(&mut self) -> anyhow::Result<Vec<(Header, Bundle)>> {
+        let this_block = self.download_block(self.current_block).await?;
+        let blob_ids = get_block_tx_blobs(&this_block, &self.target_contract)?;
+
+        let sidecars = self.get_sidecars_for_current_block().await?;
+        let mut result = Vec::new();
+        if let Some(sidecars) = sidecars {
+            result.extend(get_blobs_from_beacon_response(sidecars, blob_ids)?);
+        }
+
+        self.current_block += 1;
```

Comment:
Does it cover the use case, when bundle committed across several Ethreum blocks?

Reply:
I don't think a bundle can be committed across Ethereum blocks. The fragments are returned by this API and the higher layer should handle that. My understanding of the splitting mechanism is poor, though, so I might be mistaken here.